account for file icon

lebensterben opened the pull request:
When `dashboard-set-file-icons` is non-nil, by default it adds a unicode file icon and a space before an item on dashboard.

Before this commit, when envoking either `ace-link-dashboard` or `ace-link-dashboard-remove`, the overlay drawn by `avy` appears on top of the unicode file icon and makes it unintelligible, regardless of the value of `avy-style`.

This commit detects unicode file icons when collecting the widgets and move forword the point past the icon and the padded space. Therefore the overlay would be drawn correctly on the item.

In addition, this commit also surpresses the warning the `dashboard-remove-item-under` is not defined.

lebensterben committed:
account for file icon

When `dashboard-set-file-icons` is non-nil, by default it adds a unicode
file icon and a space before an item on dashboard.

Before this commit, when envoking either `ace-link-dashboard` or
`ace-link-dashboard-remove`, the overlay drawn by `avy` appears on top
of the unicode file icon and makes it unintelligible, regardless of the
value of `avy-style`.

This commit detects unicode file icons when collecting the widgets and
move forword the point past the icon and the padded space. Therefore the
overlay would be drawn correctly on the item.

In addition, this commit also surpresses the warning the
`dashboard-remove-item-under` is not defined.

ricardoricho commented on ace-link-dashboard.el:
```diff
@@ -75,7 +77,10 @@
       (while (< previous-point (funcall next-widget-point))
         (setq previous-point (point))
-        (push (cons (widget-at previous-point) previous-point) candidates))
+        (push (cons (widget-at previous-point) (if (eq 'unicode (char-charset (char-after (point))))
```
Could you put the `if` in a new line?
Or better, move that conditional to a new function

ricardoricho reviewed (changes requested):
Thanks

lebensterben committed:
fix `ace-link-dashboard-remove`

This commit adds an optional parameter to
`ace-link-dashboard--collect` which ensures that the widget collected is
within a known section of `dashboard-mode` buffer.

This test is done via the new function
`ace-link-dashboard--in-section-p`.

In addition, this commit also factors out the test of unicode symbol to
the new function `ace-link-dashboard--on-unicode-symbol-p`.

lebensterben commented:
Hi, please check the new commit. It fixes another bug.

Originally, when you run `ace-link-dashboard-remove`, the may include candidates outside of the known sections, such as the custom buttons users add below the banner. And it will emit a `user-error` when the user chooses to "remove" a button.

ricardoricho commented on ace-link-dashboard.el:
```diff
@@ -64,18 +67,35 @@
-        (push (cons (widget-at previous-point) previous-point) candidates))
+        (when (or (and in-section (ace-link-dashboard--in-section-p))
+                  (not in-section))
+          (push (cons (widget-at previous-point)
```
I was thinking in move the `if` to another function.
```emacs-lisp
(defun ace-link-dashboard--widget-point (point)
"Return the POINT where avy overlay will be display.
Avy don't show the correct overlay when point is over an icon"
  (if (eq 'unicode (char-charset (char-after point)))
    (+ 2 point)
    point))
```
And this line changes to:
```emacs-lisp
(push (cons (widget-at previous-point) (ace-link-dashboard--widget-point previous-point)) candidates)
```
Or something like that, to have the `push` as clean as possible.

lebensterben replied:
you can push the changes to my branch.

ricardoricho replied:
Sure.

ricardoricho reviewed (changes requested):
Hi,
Let's solve one issue at a time. I think the icon part is almost ready.

The "remove" part we can discuss it in another PR.

ricardoricho committed:
Fix move point forward when point is over an icon.

jcs090218 commented:
There is an error with the Windows snapshot, so we might want to ignore that job’s test. It might be a great idea to add `experimental` flags to all snapshots.
